### ISSUE TITLE
Negative margin causes bad pagebreak

### ISSUE THREAD

**Issue by Jellby** [koreader/koreader#10475]
* KOReader version: v2023-04-60-gd54c5acc_2023-05-21
* Device: Linux emulator (but also on Kobo Aura)

#### Issue

A float with a negative margin causes a too early pagebreak with an orphaned line, while there is more than enough space in the previous page. Please excuse the mangled text, the "W.W." at the top of the second page should be in the previous one:

![bad1](https://github.com/koreader/koreader/assets/7943459/859b0bd1-18f8-4781-935e-77e052437428)
![bad2](https://github.com/koreader/koreader/assets/7943459/2db220b9-162c-4c46-a9d1-02b6b42ce8c8)

#### Steps to reproduce

Sample epub file (rename to .epub if needed): 
[test.zip](https://github.com/koreader/koreader/files/11525288/test.zip)

The relevant CSS:

```
div.separator {
  margin: 2em 0;
  text-align: center;
}
img.seal {
  width: 5em;
  float: left;
  margin: -0.5em 0.5em 0 0;
}
div.date {
  text-align: left;
  font-style: italic;
  margin: 1em 0;*/
}
```
and I believe it's the negative margin that causes it. At least, if the top `img.seal` margin is changed to `0` and the top `div.date` margin is changed to `0.5em` (which should be equivalent in this case), the text looks good, which also proves that it fits:
![good1](https://github.com/koreader/koreader/assets/7943459/a7abbe23-8400-4959-a656-6bafb6598509)
![good2](https://github.com/koreader/koreader/assets/7943459/c812ddf6-dfd0-40a0-8073-df78caba4ebf)

No puppies are dying because of this, but maybe there's always hoping there's an easy fix :)

**Comment by poire-z** [comment]
> No puppies are dying because of this

Phew, thank you for that :)

> but there's always hoping there's an easy fix :)

Negative margins are hell (https://github.com/koreader/crengine/issues/516).
And I guess most of these cases are simpler to fix via style tweaks when they happen (ie. setting the impacted margin to 0 like you did).

It's hard to get back in my code that handles all this, but I think this may be the intended implementation goal: if there is negative margins on the float, part of it would be displayed on the previous horizontal slice/block, and so, to not cut the float content, these are considered unbreakable and brought together on the next page if they don't fit all on previous one.

Or you mean that the div.separator of 2em should be the one that the float overlaps, and so should be the only one brought with it in the next page ? Which somehow, it seems to happen when I uncomment your `border: solid 1px red` ? In which case, yes, it feels a bit more odd. But probably not easy to fix, as when no border/padding/content, these 2em margins should not be emit'ed, and just accounted in the current margins that end in a single "collapsed margins", so somehow merged with the previous P zero-margin, and still bringing them together.

**Comment by Jellby** [comment]
Yes, indeed. I expected the `div.separator` to be able to absorb any negative margin, and that the preceding paragraph would not care about whatever comes after that.

**Comment by Jellby** [comment]
Note also that although the vertical separation from the previous paragraph can made the same without using negative margins, I need the negative margin for correct vertical alignment of the square float and the text beside it, or at least it's the most obvious way.

**Comment by poire-z** [comment]
If you just need a workaround, and don't have additional top/bottom margins on the other blocks that you would expect to collapsed with your `div.separator { margin: 2em 0 }`, you could use `div.separator { padding: 2em 0 }`

**Comment by poire-z** [comment]
> but there's always hoping there's an easy fix :)

I'm afraid there is no easy fix...
Investigated, tracking how things happen:
1) (in lvtextfm.cpp) the first line of text in the paragraph (=final block) with the float, as its float has some negative margin, gets a "page break before avoid" flag
2) upper on in lvrend.cpp, when processing blocks, and then each line of "final blocks" to add them to the vertical roll, we get that flag, but we don't and can't know the reason
3) we then emit that first line - which causes the accumulated vertical margins to be pushed onto the vertical roll. At the point of doing that, we don't know the sources/parts of the accumulated margins: it's now a single "collapsed margin". Margins forward the flags of their non-margin neighbours (if a first line of text has a real `page-break-before: avoid`, it is forwarded across the above margin so the last line of the previous paragraph doesn't break (from the margin, and so from the first line of the next paragraph).

To properly handle your case, we would need to 1) gather that the "page break before avoid" flag is from an ongoing float, and remember by how much pixels it is needed. 2) handle that in the accumulated margins, so we know we can split it safely, and where, while 3) still handling proper break avoid flags.
So, quite a bit too complex to take the risk of breaking this already complex and fragile handling :/

As the author/publisher, if you can find a workaround, please let me know how you ended up doing it.

**Comment by Jellby** [comment]
> (in lvtextfm.cpp) the first line of text in the paragraph (=final block) with the float, as its float has some negative margin, gets a "page break before avoid" flag

Hmm... Is that per-spec or is it a (reasonable) tweak? Is there any way it could be disabled (custom css, user setting...)?

**Comment by Jellby** [comment]
Otherwise, the best I can come up with so far is adding a fake element after the float that takes care of the vertical offset (instead of negative margin):

```
<div class="separator"></div>

<div class="date">
<img class="seal" src="images/seal.png" alt=""/><span/>
W Woooooo oo Woooooo, W.W.W., oo Wooo, W.W.W.
</div>
```
```
img.seal {
  width: 5em;
  float: left;
  margin: 0em 0.5em 0 0;
}
img.seal + span {
  display: block;
  height: 0.5em;
}
```

I tried with `<br/>` instead of `<span/>` (at least it's closer to the intent), but although it works in Firefox, it doesn't in KOreader or the calibre viewer: the `<br/>` moves a whole line down, not just `0.5em`.

**Comment by poire-z** [comment]
> > (in lvtextfm.cpp) the first line of text in the paragraph (=final block) with the float, as its float has some negative margin, gets a "page break before avoid" flag
> 
> Hmm... Is that per-spec or is it a (reasonable) tweak? Is there any way it could be disabled (custom css, user setting...)?

Not per specs, but some implementation "trick" (given all the various code involved, text, blocks, page splitting - each step being not really aware of what's happening in the others) to ensure a float is not broken across pages.

> Otherwise, the best I can come up with so far is adding a fake element after the float that takes care of the vertical offset (instead of negative margin):

Can't you just give some "body" to your div.separator, so it's not pure margins: add it some `padding-top: 1px` and may be just some margin-bottom for the float ? Any padding/border should be enough to "split" its margin and stop the "break: avoid" from propagating?

**Comment by Jellby** [comment]
But I want it to collapse with other margins and at (automatic) page break.

**Comment by poire-z** [comment]
What "it" ?

![image](https://github.com/koreader/koreader/assets/24273478/5373f319-2339-46b7-9100-50b1438437e1)

I get that the blue is your DIV.date top, and the red is what you need to have your float overflow (for correct aligment with the DIV.date text).
We won't cut a page at the blue, because it would cut your float (and no, we won't allow it to be in the top page margin for free :) although I remember there are cases where I did stuff so glyphs could overflow in the top margin, but may be it's just for glyphs, when they overflow the line height).
So, you just need to add any margin above the red, and it will collapse with any previous margin (if any, there doesn't seem to any in your sample, or may be I messed up your sample).
It looks to me like this does the trick (for this particular bad page break):
```css
div.separator {
  text-align: center;
  padding-top: 1px; margin-bottom: 0.5em; /* added */
              /* add any margin-top here you may wish */
  /*border: solid 1px red;*/
}
img.seal { /* untouched */
  width: 5em;
  float: left;
  margin: -0.5em 0.5em 0 0;
  /*margin: 0em 0.5em 0 0;*/
}

div.date { /* don't remember if I touched that :) */
  text-align: left;
  font-style: italic;
  margin: 1em 0;
  /*margin: 0.5em 0 1em 0;*/
}
```

**Comment by poire-z** [comment]
> I tried with `<br/>` instead of `<span/>` (at least it's closer to the intent), but although it works in Firefox, it doesn't in KOreader or the calibre viewer: the `<br/>` moves a whole line down, not just `0.5em`.

Yes, a `<br/>`, per specs or the way it is implemented, not sure :), moves by the paragraph "strut" height, that is the height of a line box implied by the line-height set on the paragraph.
Your span acts as a inline-block, so it's an okay trick to shift the baseline down and ensure the first line gets a taller height so the float it fully inside the DIV.data - and the margin business above is not being messed up by the float.
